Introduce JwtEncoder

jgrandja opened the pull request:
A `JwtEncoder` encodes the JOSE headers and JWT Claims Set into a JSON Web Token (JWT).

The initial implementation `NimbusJwtEncoder` supports encoding a JWS (JSON Web Signature).

jzheaux commented on crypto/src/main/java/org/springframework/security/crypto/key/CryptoKey.java:
```diff
@@ -0,0 +1,234 @@
+ * @since 5.5
+ * @see CryptoKeySource
+ */
+public class CryptoKey<K extends Key> {
```
The name concerns me a bit here since "Key" carries some semantic weight in JCA/JCE. Since it's not a key itself, I think its name should instead reflect something that contains a key.

jzheaux commented on crypto/src/main/java/org/springframework/security/crypto/key/CryptoKey.java:
```diff
@@ -0,0 +1,234 @@
+
+	private final K key;
+
+	private final String id;
```
This wouldn't work very well for SAML, which doesn't have a notion of selecting credentials by id. I wonder if it would be more appropriate living in `oauth2-jose` and being called `Jwk`.

jzheaux commented on oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jose/jws/NimbusJwsEncoder.java:
```diff
@@ -0,0 +1,340 @@
+	private static final String EC_KEY_TYPE = "EC";
+
+	private static final Map<JwsAlgorithm, String> jcaKeyAlgorithmMappings = new HashMap<JwsAlgorithm, String>() {
+		{
```
A quick search through the code base shows that DBI is pretty uncommon. It's also listed as [an anti-pattern](https://rules.sonarsource.com/java/RSPEC-3599) by a few different sources.

I think it's more reasonable to place this in the constructor.

jzheaux commented on oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/JwtClaimsSet.java:
```diff
@@ -0,0 +1,191 @@
+	 * Returns a new {@link Builder}.
+	 * @return the {@link Builder}
+	 */
+	public static Builder withClaims() {
```
I would have expected this method to take `claims` as a parameter. Should it be called `builder()` instead, since it doesn't take any parameters?

jzheaux commented on crypto/src/main/java/org/springframework/security/crypto/key/SymmetricKey.java:
```diff
@@ -0,0 +1,62 @@
+
+/**
+ * A {@link CryptoKey} that holds a {@code javax.crypto.SecretKey} used for symmetric
+ * algorithm's.
```
Not to take away from the more important discussion about contracts, but it would be clearer to readers if the apostrophe were only used when talking about possession.

Here, for example, `used for symmetric algorithms`, is correct. There are a few other places in the PR where the possessive ' is being used to pluralize.

jzheaux reviewed (commented):
Nice, @jgrandja! I've left some feedback inline.

rwinch reviewed (changes requested):
Can you help me understand why this code needs to go into Spring Security

jgrandja commented:
Thanks for the feedback @jzheaux @rwinch.

I decided to remove `CryptoKey` and `CryptoKeySource` in favour of `KeySource`, which "supplies" `java.security.KeyPair`'s and/or `javax.crypto.SecretKey`'s.

I introduced `JwtEncoderFactory` (aligned with existing `JwtDecoderFactory`) with initial implementation `NimbusJwsEncoderFactory` that creates `NimbusJwsEncoder`.

The `NimbusJwsEncoderFactory` requires a `Supplier<JWKSet>`, which could be defined with the help of `NimbusKeySourceJWKSetConverter` and the application-provided `KeySource`.

jzheaux reviewed (changes requested):
Nice, @jgrandja! I've left some feedback inline.

jgrandja commented on oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/NimbusJwsEncoderFactory.java:
```diff
@@ -0,0 +1,99 @@
+						"Found multiple signing keys for algorithm '" + jwsAlgorithm.getName() + "'"));
+			}
+
+			return new NimbusJwsEncoder(jwks.get(0));
```
Good catch! I'll update to support a key selection strategy.

jgrandja commented on oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/NimbusJwsEncoder.java:
```diff
@@ -0,0 +1,266 @@
+	}
+
+	@Override
+	public Jwt encode(JoseHeader headers, JwtClaimsSet claims) throws JwtEncodingException {
```
Agreed. Will fix.

jgrandja commented:
@jzheaux I updated the PR to support key rotation via a `JWK` selection strategy. Take a look at this [test](https://github.com/spring-projects/spring-security/pull/9208/files#diff-1d6139dc43e98e3d8a7b19eccedadfe0cb1a91700f746cc5a8438a6969b58935R228), which demonstrates how an application would configure.

jzheaux reviewed (changes requested):
Thanks for the updates, @jgrandja. I've left some more feedback inline.

rwinch reviewed (commented):
I've left some comments inline.

A general bit of feedback is that I think we need to make some changes to avoid classes that have interface based methods with inputs/outputs that belong to external libraries (i.e. Nimbus). An example of this is `NimbusKeySourceJWKSetConverter` which has a return value of `JWKSet` (part of Nimbus) on an interface method.

When an interface method accepts/returns a value that belongs to an external library that interface cannot be used by anything except by another API that uses Nimbus. In scenarios where this is absolutely required, we typically nest the class so it is clear it belongs to a specific implementation.

jgrandja commented on crypto/src/main/java/org/springframework/security/crypto/key/KeySource.java:
```diff
@@ -0,0 +1,52 @@
+ * @see SecretKey
+ */
+@FunctionalInterface
+public interface KeySource {
```
Agreed. The recent improvements made `KeySource` unnecessary at this point, as well as `NimbusKeySourceJWKSetConverter`. I'll remove both.

jgrandja commented on oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/NimbusJwsEncoder.java:
```diff
@@ -0,0 +1,341 @@
+	 * @param jwtCustomizer the {@link Jwt} customizer to be provided the
+	 * {@link JoseHeader.Builder} and {@link JwtClaimsSet.Builder}
+	 */
+	public void setJwtCustomizer(BiConsumer<JoseHeader.Builder, JwtClaimsSet.Builder> jwtCustomizer) {
```
> It leaks the fact that the implementation is doing an internal copy of the headers and claims

At a minimum, the `JwtEncoder` needs to enhance the `JoseHeader` with some metadata of the selected key, for example `kid`. Therefore, it needs to make a copy of it so it can add the specific JOSE headers. Furthermore, this customizer allows an application to add custom claims and/or headers.

jgrandja replied:
FYI, I have considered updating `JwtEncoder` to:

`Jwt encode(JoseHeader.Builder headersBuilder, JwtClaimsSet.Builder claimsBuilder)`

but it didn't feel right at the time. I'd like to get your thoughts on this @jzheaux @rwinch

jgrandja commented on oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/NimbusJwsEncoder.java:
```diff
@@ -0,0 +1,341 @@
+	 * Constructs a {@code NimbusJwsEncoder} using the provided parameters.
+	 * @param jwkSelector the {@code com.nimbusds.jose.jwk.JWK} selector
+	 */
+	public NimbusJwsEncoder(Function<JoseHeader, JWK> jwkSelector) {
```
Agreed. I'll replace `Function<JoseHeader, JWK>` with `JWKSource`

rwinch replied:
Marking as resolved as it is complete.

jgrandja commented on oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/NimbusJwsEncoder.java:
```diff
@@ -0,0 +1,341 @@
+			JWSHeader jwsHeader = new JWSHeader(jwsAlgorithm);
+			JWKSelector jwkSelector = new JWKSelector(JWKMatcher.forJWSHeader(jwsHeader));
+			List<JWK> jwks = jwkSelector.select(this.jwkSetProvider.get());
+			if (jwks.size() > 1) {
```
> I'm assuming that the JWK source here is the same JWK source that an application would use for their `/jwks` authorization server endpoint

No. The JWK Source for the Authorization Server would behave differently than the one with Resource Server.

When a JWT needs to be signed, the JWK Source must return one `JWK` for signing. If it returns multiple, then how does it choose which one to use for signing? There is no metadata associated to the JWK indicating the "active" key. **NOTE:** This was the original thinking behind the design of the `ManagedKey`, as it would hold this metadata attribute indicating the active flag. Now that we removed `ManagedKey`, the JWK Source must return the one `active` (current) key used for signing.

The resource server could contain 2 keys that are active but it will narrow the selection process using a unique JOSE header, e.g. `kid`

scrocquesel replied:
@jgrandja I agree with you, there can only be one signing key at a time. Is it possible to not require the caller code to provide the kid or alg in the header ? This could be guess from the signing key source if provided and will avoid the caller code to have access to the jwksource, parse it and extract header value which is already done by the encoder.

I think in most use case, it is the key that drive the kid and alg set in the header. The caller code just want to encode a jwt and don't bother which alg or key will be used. It is often generated outside and rolling. If the algorithm change, the caller code need to change if it has been hard coded like the sample given above `encoder.encode(Jwt.withAlgorithm(JwsAlgorithm.RS256).subject("subject"))`

jgrandja replied:
@scrocquesel NOTE: The code in this PR is out-of-date and needs to be synced with the code currently in the package [org.springframework.security.oauth2.client.endpoint](https://github.com/spring-projects/spring-security/tree/main/oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/endpoint). The implementation `NimbusJwsEncoder` is used for Jwt Client Authentication introduced in `5.5`.

You will notice that all classes are `package-private` as we were not ready to expose `JwtEncoder` in `5.5` but we needed to add the Jwt Client Authentication feature, hence the reason for `package-private`.

If you have a need for the `JwtEncoder` API, it would be helpful if you could detail your requirements and we could consider exposing it in `5.6`

scrocquesel replied:
Thanks, I didn't notice it was old code. Should I express needs in this issue ?

jgrandja replied:
Yes. Please add a new comment at bottom of this PR detailing your requirements and the need for the `JwtEncoder` API. The more detail the better. Thanks.

jgrandja commented on crypto/src/main/java/org/springframework/security/crypto/key/KeySource.java:
```diff
@@ -0,0 +1,52 @@
+	 * @return a {@code Set} of {@code javax.crypto.SecretKey}(s), or an empty {@code Set}
+	 * if not available
+	 */
+	default Set<SecretKey> getSecretKeys() {
```
`KeySource` will be removed, see [comment](https://github.com/spring-projects/spring-security/pull/9208#discussion_r556453050)

jgrandja commented on oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jose/NimbusKeySourceJWKSetConverter.java:
```diff
@@ -0,0 +1,104 @@
+ * @see KeySource
+ * @see com.nimbusds.jose.jwk.JWKSet
+ */
+public final class NimbusKeySourceJWKSetConverter implements Converter<KeySource, JWKSet> {
```
`NimbusKeySourceJWKSetConverter ` will be removed, see [comment](https://github.com/spring-projects/spring-security/pull/9208#discussion_r556453050)

jgrandja commented on oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/JoseHeader.java:
```diff
@@ -0,0 +1,351 @@
+	 * the JWE.
+	 * @return the JWK Set URL
+	 */
+	public String getJwkSetUri() {
```
The general pattern we've been using for `URL` is that the builder would take in a `String` and the accessor would return a `URL`. For example:

```java
OidcIdToken.Builder.issuer(String issuer)

URL OidcIdToken.getIssuer()
```

This makes it convenient for the caller not having to create a `URL` when passing to the builder. The accessor returns the appropriate type via conversion, which occurs on write.

**NOTE:** The reason I decided to return `String` for `getJwkSetUri()` is that it's a `String` representation in the JOSE header anyway. What are your thoughts? Should I change it to `URL getJwkSetUri()`?

jgrandja commented on oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jose/NimbusKeySourceJWKSetConverter.java:
```diff
@@ -0,0 +1,104 @@
+		Assert.notNull(keySource, "keySource cannot be null");
+
+		// @formatter:off
+		List<JWK> jwks = keySource.getKeyPairs().stream()
```
Yes, forgot about that. Either way, `NimbusKeySourceJWKSetConverter` will be removed.

jgrandja commented on oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/JoseHeader.java:
```diff
@@ -0,0 +1,351 @@
+	private final Map<String, Object> headers;
+
+	private JoseHeader(Map<String, Object> headers) {
+		this.headers = Collections.unmodifiableMap(new LinkedHashMap<>(headers));
```
I'll change it to `HashMap`

jgrandja commented on oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/JoseHeader.java:
```diff
@@ -0,0 +1,351 @@
+	 * the JWS or encrypt the JWE.
+	 * @return the X.509 URL
+	 */
+	public String getX509Uri() {
```
See [comment](https://github.com/spring-projects/spring-security/pull/9208#discussion_r556480993)

jgrandja commented on oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/JwtEncoder.java:
```diff
@@ -0,0 +1,59 @@
+	 * @return a {@link Jwt}
+	 * @throws JwtEncodingException if an error occurs while attempting to encode the JWT
+	 */
+	Jwt encode(JoseHeader headers, JwtClaimsSet claims) throws JwtEncodingException;
```
I'm in favour of keeping the operation as-is or consider changing it to:

`Jwt encode(JoseHeader.Builder headersBuilder, JwtClaimsSet.Builder claimsBuilder)`

Introducing the [JoseHeader](https://tools.ietf.org/html/rfc7519#section-5) and [JwtClaimsSet](https://tools.ietf.org/html/rfc7519#section-4) constructs aligns with the spec.

The `Jwt.Builder` currently supports setting the typed and custom claims but the headers only generically via `header()` or `headers()`. Adding all the possible JWS and JWE headers to `Builder` is going to make the `Builder` quite large and not as easy to understand without the clear separation of claims and headers, which the new constructs would achieve.

I would even prefer to re-work `Jwt`, specifically the `Builder` and associated factory method `withTokenValue(String tokenValue)`. For example, some of the proposed changes would be:

```java
`JoseHeader Jwt.getHeaders()`

`JwtClaimsSet Jwt.getClaims()`

static Jwt.with(JoseHeader, JwtClaimsSet)
```

I know what I'm proposing as far as the re-work for `Jwt` is a big change but I strongly feel we should have separate representations of the JOSE headers and Jwt Claims set. I believe this will keep things simpler (in each class), easier to maintain and easier to grow when we introduce support for JWE at a later point (which adds new headers to the mix).

@jzheaux @rwinch What are your thoughts on this?

jzheaux replied:
> For the Nested JWT use case, the returned object cannot be a `Jwt`.


I don't really see why this is the case, could you elaborate? Couldn't `Jwt` contain three members, like so:

```java
JwtClaimsSet getClaimsSet();
JwsHeader getJwsHeader();
JweHeader getJweHeader();
```

Flattening a structure into its relevant parts like this is a practical way for an application to consume it. It's uncommon for folks to need to know the headers anyway -- most callers only need to know the resulting encoded value and its claims.

I imagine that a change like this will have to happen anyway to `Jwt` so that `Jwt NimbusJwtDecoder#decode(String)` can decrypt and verify an encrypted and signed JWT.

If `Jwt` cannot be made to suit, I wonder if this means that `JwtEncoder#encode`'s return type needs to be different from `Jwt`. I believe returning a `String` can address this concern.

Practically speaking, returning a serialized value is less of a risk than it may seem on the surface. In all cases, the claims must be specified by the caller, so the caller does not lose access to the claims when a string is returned. I believe it will be uncommon for the caller to need to know the headers that were used -- but if this is needed, an application can easily parse the return value with something like `JWTParser.parse` from Nimbus.

The use case we currently have in the codebase, `NimbusJwtClientAuthenticationParametersConverter`, only needs the serialized value. A brief Find Usages search in Spring Authorization Server shows that only the serialized value and the claims are needed to fulfill those use cases.

If `Jwt` cannot be made to suit and if `String` feels too risky, I wonder if we could introduce `EncodedJwt` and return that. It could initially contain only the encoded `String`. Additional members could be added if needed.

I'm keen on making this aspect of the contract work. Given that `JwtDecoder` has the potential to decode both signed JWTs and Nested JWTs because of its contract, I think it's reasonable to see if `JwtEncoder` can encode both signed JWTs and Nested JWTs with the right contract. Similar to applications asking Spring Security to "please decrypt and verify this" without exposing them to Nested JWT internals (Nimbus does exactly this), it seems quite reasonable to give Spring Security a set of claims and say "please sign and encrypt this". 

Parenthetically, if we can't do this, I think it will be less confusing if the interface is called `JwsEncoder` so that users don't think to align the capabilities of `JwtDecoder` with `JwtEncoder`.

rwinch replied:
I'm marking this as resolved because the code now uses JwtEncoderParameters. We can add any additional feedback on the updated code.

jgrandja commented on etc/checkstyle/checkstyle-suppressions.xml:
```diff
@@ -17,6 +17,7 @@
 
 	<!-- InterfaceIsType rules we can't fix until a major revision due to back compatibility -->
 	<suppress files="JwsAlgorithms\.java" checks="InterfaceIsType"/>
+	<suppress files="JoseHeaderNames\.java" checks="InterfaceIsType"/>
```
For sure. I'll change the new classes to be `final class` or `abstract class`. Do you have a preference from the 2 options? I'll also log a ticket to re-factor the existing interface types.

jgrandja commented on oauth2/oauth2-jose/src/test/java/org/springframework/security/oauth2/jwt/NimbusJweEncoderTests.java:
```diff
@@ -0,0 +1,484 @@
+
+	// @formatter:off
+	/*
+	 * IMPORTANT DESIGN DECISION
```
@jzheaux FYI

rwinch replied:
@jzheaux @jgrandja Does this still need attention?

jgrandja replied:
@rwinch I'm currently working on some updates that may very well resolve this. Let's keep it for now until I complete updates.

scrocquesel commented:
Signing keys are often managed externally to the application and alg may change when rolling out new keys.
As caller code, I don't want to set headers related to signing key (alg, kid), so that I don't have to fiddle with the content of the keys.

Actually, the Jwt given to encoder must provide a alg header. Either the caller code reads the alg from a configured property (often it is hard coded) or it has access to the same JwkSource and extracts the alg by itself (as seen [here](https://github.com/spring-projects/spring-security/blob/68f91edbb8a223ee8505893697d85cbe51890ea2/oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/endpoint/NimbusJwtClientAuthenticationParametersConverter.java#L116)). When there is only one signing key in the JwkSource, kid and alg should be guess from this signing key.

Note that this require alg and kid members in JWK but JWK RFC 7517 set them as optional. This shouldn't be a new requirement as currently, the encoder seems to require them both already.

jgrandja commented:
`JwtEncoder` will be required for gh-10232. It MAY also be required for gh-9880.

**NOTE:** `JwtEncoder` and related classes in this PR is currently used in Spring Authorization Server. Merging this here will allow for reuse.

rwinch reviewed (changes requested):
I think we just need to finalize our thoughts around the arguments to the encode method.

jzheaux reviewed (changes requested):
I've left three separate comments that are related to the method parameters in the contract @rwinch indicated.

Additionally, I recently noticed a concern about the `crit` header, and so I've added some details about that too.

jgrandja commented:
@jzheaux Regarding your comment...

> The nice thing about a parameters object is that it's also then possible to introduce something like the following:

> `Jwt signedAndEncryptedJwt = encoder.encode(JwtEncoderParameters.signThenEncrypt(claims));`

For the Nested JWT use case, the returned object cannot be a `Jwt`. The example provided `signThenEncrypt` is a JOSE object structure that has a JOSE header and the (nested) payload is a signed JWS, which also contains a JOSE header and the claims set. Therefore, the return value as per example above **cannot be** a `Jwt` because that payload is a claims set only.

> without changing the contract. As I understand the current proposal (please expand NimbusJweEncoderTests when following the link), this particular use case would be several lines of code and would require additional interfaces.

Hence the additional interface/classes in `NimbusJweEncoderTests`.

jgrandja commented:
@jzheaux I looked at the Nimbus API `JWSHeader` and `alg` is not being defaulted to `RS256`. I think you mentioned it was defaulted but I may have misheard. Can you confirm?

Also, I checked the Nimbus API `JWEHeader` and neither `alg` or `enc` are being defaulted.

Maybe you're seeing something that I'm not seeing?

jgrandja commented:
@rwinch @jzheaux I'm having second thoughts on introducing `JwtEncoderParameters`.

From my understanding in our discussion, the main goal of `JwtEncoderParameters` is to solve:

> I'd like to see if we can find an encode contract where it's easier to supply headers optionally

We decided that we would default to `RS256` if headers are not supplied. Therefore another solution can be:

```java
public final class JoseHeaders {

	public static final JoseHeader RS256_ALG = JoseHeader.withAlgorithm(SignatureAlgorithm.RS256).build();

	private JoseHeaders() {
	}

}
```

With usage:

```
JwtClaimsSet jwtClaimsSet = ...
Jwt encodedJwt = jwtEncoder.encode(JoseHeaders.RS256_ALG, jwtClaimsSet);
```

This amounts to the same lines of code as the proposed solution for `JwtEncoderParameters`.

It still introduces a new class `JoseHeaders` but it's much simpler as it only defines constants. This can be an `enum` instead.

I'd like to understand what are the other benefits that `JwtEncoderParameters` can provide other than providing a default for `alg`? Can you list some other benefits @rwinch @jzheaux ?

I really want to keep things simple and not introduce any extra classes that are not needed.

IMO, the extra boilerplate code `JoseHeader header = JoseHeader.withAlgorithm(SignatureAlgorithm.RS256).build()` should not be the only benefit that `JwtEncoderParameters` provides.

jzheaux commented:
> I think you mentioned it was defaulted but I may have misheard. Can you confirm?

My apologies, I don't remember saying that. Part of the confusion might have been my overloading the term "implementation". I meant it in the OOP sense: The implementation (`NimbusJwtEncoder`) would figure out the algorithm in the event that one is not specified by the caller. 

I imagine this obligation to figure out the algorithm could be handled by the underlying library, though something like that won't affect `JwtEncoder#encode`'s contract.

> We decided that we would default to RS256 if headers are not supplied.

Yes, that's my recollection, too. I would recommend this defaulting be done by `NimbusJwtEncoder`, though, not by the method parameter. The reason is that having the implementation do it allows implementations to configure different defaults other than `RS256`. 

And while a prototype `JoseHeader.Builder` could hypothetically be exposed with a different algorithm specified, I think introducing a setter into `NimbusJwtEncoder` would be a more familiar pattern for developers. It's also nice because it aligns with how algorithms are configured in `NimbusJwtDecoder`.

jzheaux commented on oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/JwtEncoderParameters.java:
```diff
@@ -0,0 +1,68 @@
+	 * Returns the {@link JoseHeader headers}.
+	 * @return the {@link JoseHeader}
+	 */
+	public JoseHeader getHeaders() {
```
Thanks for looking into this. I think my motivations here are tied to wanting a contract that can support Nested JWTs, so I will focus my feedback along those lines.

To be clear, I am imagining that `JwtEncoderParameters` could eventually look like the following (assuming renaming of classes):

```java
JwtClaimsSet getClaims();
JwsHeader getJwsHeader();
JweHeader getJweHeader();
```

This, of course, makes sense in the context of Nested JWTs -- for the contract to support sign-then-encrypt, it would be important to be able to supply both signing headers and encryption headers in the method call.

> I suppose we could have `getJwsHeader()` and `getJweHeader()` but that would not be a clean design from my perspective.

Can you elaborate on why this isn't a clean design from your perspective? I may be able to help if I can understand the specific issue.

jgrandja replied:
> Can you elaborate on why this isn't a clean design from your perspective?

A JWT is either a JWS or a JWE, not both. So having `getJwsHeader()` and `getJweHeader()` does not make sense from a design standpoint. Even in the Nested JWT use case (sign then encrypt), it's still a JWE but the payload is a JWS.

Is the user to assume that `getJwsHeader()` is dual purpose? The header for a JWS OR the header for a nested JWS?

jzheaux replied:
> A JWT is either a JWS or a JWE, not both

I agree with this, though I'm not seeing the connection to `JwtEncoderParameters`. This class is not trying to represent a JWT but instead the parameters needed to encode a JWT. It may even contain methods and fields that are not headers or claims.

> Is the user to assume that `getJwsHeader()` is dual purpose? 

The user need not assume anything since they are using factory methods to describe their intent.

The implementation then reads the parameters object to infer the user's intent. If the parameters object indicates that signing should be done, then the implementation does so. After that, if the parameters object also indicates that encryption should be done, then the implementation does that, too.

jzheaux replied:
I think we could possibly be going afield from [the main reason](https://github.com/spring-projects/spring-security/pull/9208#discussion_r710349600) (see the second-to-last paragraph) for my suggestion, though. We can certainly make anything work - we want to make sure that we are building the right thing.

IMO, `JwtDecoder` and `JwtEncoder` should have comparable capabilities. [`JwtDecoder`'s contract supports a Nested JWT](https://github.com/spring-projects/spring-security-samples/blob/main/servlet/spring-boot/java/oauth2/resource-server/jwe/src/main/java/example/OAuth2ResourceServerSecurityConfiguration.java#L88) without additional interfaces or method calls, and so it seems that `JwtEncoder` should be able to as well.

rwinch commented:
I think @jzheaux already highlighted why I prefer an argument that contains the parameters, but to be explicit it provides us with the option of explicitly providing the headers, defaulted by the headers, or by the `JwtEncoder` implementation.

jgrandja commented:
@jzheaux @rwinch Thanks for all the feedback! I'm quite happy with the improvements. This is now ready for a full review. I believe I have addressed all your concerns @jzheaux.

**NOTE:** You will notice a commit _"Disable JWE tests"_ , which simply comments out some code in `NimbusJweEncoderTests` so the build passes. I will work on this next and update it to use the new design. This should not stop you from doing a review at this point. I'll ping you both when that has been updated.

jgrandja commented:
@jzheaux @rwinch I just pushed the updates for `NimbusJweEncoderTests`. The great news is that the new interface/classes are no longer needed with the new design of `JwtEncoder`. The tests prove that the current design meets the requirements of JWS, JWE, and JWE+NestedJWS.

@jzheaux Your feedback was very valuable. Thank you!

jzheaux commented on oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/NimbusJwtEncoder.java:
```diff
@@ -89,28 +74,36 @@
 	 * @param jwkSource the {@code com.nimbusds.jose.jwk.source.JWKSource}
 	 */
-	NimbusJwsEncoder(JWKSource<SecurityContext> jwkSource) {
+	public NimbusJwtEncoder(JWKSource<SecurityContext> jwkSource) {
```
When support for Nested JWTs is added, you will need a constructor that takes two `JWKSource` values, one for this application's signing keys and one for the recipient's encryption keys. 

Given that, I don't think it's an issue to instead take Nimbus's `JWSMinter` as a parameter instead of the `JWKSource`. I like taking a `JWSMinter` better because it makes the intent of the constructor unambiguous, that these keys are for signing, not for encrypting.

Alternatively, you could use a static factory method, e.g. `fromSigningJwkSource(jwkSource)` or similar.

jgrandja replied:
When we add support for JWE, we can add another constructor that either takes 2x `JWKSource` or `JWSMinter` and `JWKSource`.

`JWKSource` is currently used throughout the code but there is no usage of `JWSMinter`.

Do you see any issues if we stick with `JWKSource`?

jzheaux replied:
> Do you see any issues if we stick with `JWKSource`?

One `JWKSource` seems ambiguous. If `NimbusJwtEncoder` supports encryption as well, is the single-`JWKSource` constructor for signing or for encryption?

> `JWKSource` is currently used throughout the code but there is no usage of `JWSMinter`.

I think `JWSMinter` makes more sense as a dependent component, similar to `JWTProcessor` on the `JwtDecoder` side. It would imply changing the implementation to use `JWSMinter`.

If "un-nested" JWE is not a use-case or if the ambiguity caused by a single-`JWKSource` constructor is not a concern, then the implementation could add a `JWSMinter` constructor later on, though I don't really see a reason to wait unless the Nimbus component is not yet suitable.

jzheaux commented on oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/JwtEncoderParameters.java:
```diff
@@ -0,0 +1,81 @@
+	 * @param claims the {@link JwtClaimsSet}
+	 * @return the {@link JwtEncoderParameters}
+	 */
+	public static JwtEncoderParameters with(JwtClaimsSet claims) {
```
I'm thinking that `with` is usually for builders and `from` is for static factory methods. Am I mistaken? If so, please disregard.

jgrandja replied:
Changed to `from()`

jzheaux reviewed (changes requested):
Thanks, @jgrandja, for the updates! I've left some additional feedback inline about the changes.

jgrandja commented on oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/JoseHeader.java:
```diff
@@ -310,72 +326,80 @@ Builder x509SHA1Thumbprint(String x509SHA1Thumbprint) {
+
+		@SuppressWarnings("unchecked")
+		protected void validate() {
+			Set<String> criticalHeaderNames = (Set<String>) this.headers.get(JoseHeaderNames.CRIT);
```
As per [spec](https://datatracker.ietf.org/doc/html/rfc7515#section-4.1.11):

> Its value is an array listing the Header
   Parameter names present in the JOSE Header that use those extension

Hence the type should be `Set<String>` for the `crit` header to align with the spec. The suggestion to change to `criticalHeaders(Map<String, Object)` wouldn't align with the spec. We could offer this **additional** mutator in the builder as a convenience method but I don't believe it's needed at the moment. IMO the `crit` header is an advanced header and I believe the user would be aware of setting the actual custom headers before setting the `crit` header.

jgrandja replied:
@jzheaux As discussed, I added `criticalHeader(String name, Object value)` and `criticalHeaders(Consumer<Map<String, Object>> headersConsumer)`. The `Consumer` method adds a lot of extra code and some internal complexity. It also exposes additional `public` and `protected` methods.

As an exercise, I removed `criticalHeaders(Consumer<Map<String, Object>> headersConsumer)` and it greatly simplified the code. The `criticalHeader(String name, Object value)` method still provides the convenience that we're looking for and greatly simplifies the code. My preference would be to only add this one method. See this commit for changes 132f833031bc8a6c3f24a5fa316a1f8d8b0e6bf0

jgrandja commented on oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/JoseHeader.java:
```diff
@@ -60,11 +44,12 @@
  * Header</a>
  */
-final class JoseHeader {
+public class JoseHeader {
```
I made `JoseHeader` `package-private` instead. This will make it easier when we add support for JWE.

jgrandja commented on oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/JwtEncoderParameters.java:
```diff
@@ -0,0 +1,81 @@
+	 * @param claims the {@link JwtClaimsSet}
+	 * @return the {@link JwtEncoderParameters}
+	 */
+	public static JwtEncoderParameters with(@Nullable JwsHeader jwsHeader, JwtClaimsSet claims) {
```
Done
